### ISSUE TITLE
Make fails on M1 mac due to unsupported -march=native flag

### ISSUE THREAD

**Issue by jwyles-ahana** [facebookincubator/velox#1328]
Running make after successfully running scripts/setup-macos.sh results in the following error message on M1 mac.

```
[2/859] Building CXX object velox/experimental/codegen/ast/CMakeFiles/velox_codegen_ast.dir/UnaryExpressions.cpp.o
FAILED: velox/experimental/codegen/ast/CMakeFiles/velox_codegen_ast.dir/UnaryExpressions.cpp.o 
ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DCODEGEN_ENABLED=1 -I/Users/jwyles/code/velox1/velox/. -I/Users/jwyles/code/velox1/velox/third_party/googletest/googletest/include -I/Users/jwyles/code/velox1/velox/third_party/googletest/googlemock/include -isystem /opt/homebrew/include -isystem /Users/jwyles/code/velox1/velox/velox -isystem /Users/jwyles/code/velox1/velox/velox/external -isystem /Users/jwyles/code/velox1/velox/velox/external/duckdb -mavx2 -mfma -mavx -mf16c -mbmi2 -march=native -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Werror -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-range-loop-analysis -Wno-mismatched-tags -O3 -DNDEBUG -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk -mmacosx-version-min=12.2 -std=gnu++17 -MD -MT velox/experimental/codegen/ast/CMakeFiles/velox_codegen_ast.dir/UnaryExpressions.cpp.o -MF velox/experimental/codegen/ast/CMakeFiles/velox_codegen_ast.dir/UnaryExpressions.cpp.o.d -o velox/experimental/codegen/ast/CMakeFiles/velox_codegen_ast.dir/UnaryExpressions.cpp.o -c /Users/jwyles/code/velox1/velox/velox/experimental/codegen/ast/UnaryExpressions.cpp
clang: error: the clang compiler does not support '-march=native'
```

Looks like this flag is not supported on M1 processors: https://stackoverflow.com/questions/65966969/why-does-march-native-not-work-on-apple-m1

**Comment by yingsu00** [comment]
Related issue: https://github.com/facebookincubator/velox/issues/833

**Comment by jwyles-ahana** [comment]
After removing the SIMD related flags we got this error:

```
[10/858] Building CXX object velox/duckdb/conversion/CMakeFiles/velox_duckdb_conversion.dir/DuckWrapper.cpp.o
FAILED: velox/duckdb/conversion/CMakeFiles/velox_duckdb_conversion.dir/DuckWrapper.cpp.o 
ccache /Library/Developer/CommandLineTools/usr/bin/c++ -DBOOST_ATOMIC_DYN_LINK -DBOOST_ATOMIC_NO_LIB -DBOOST_CONTEXT_DYN_LINK -DBOOST_CONTEXT_NO_LIB -DBOOST_FILESYSTEM_DYN_LINK -DBOOST_FILESYSTEM_NO_LIB -DBOOST_PROGRAM_OPTIONS_DYN_LINK -DBOOST_PROGRAM_OPTIONS_NO_LIB -DBOOST_REGEX_DYN_LINK -DBOOST_REGEX_NO_LIB -DBOOST_SYSTEM_DYN_LINK -DBOOST_SYSTEM_NO_LIB -DBOOST_THREAD_DYN_LINK -DBOOST_THREAD_NO_LIB -DCODEGEN_ENABLED=1 -DFMT_LOCALE -DGFLAGS_IS_A_DLL=0 -I/Users/jwyles/code/velox1/velox/. -I/Users/jwyles/code/velox1/velox/third_party/googletest/googletest/include -I/Users/jwyles/code/velox1/velox/third_party/googletest/googlemock/include -isystem /opt/homebrew/include -isystem /Users/jwyles/code/velox1/velox/velox -isystem /Users/jwyles/code/velox1/velox/velox/external -isystem /Users/jwyles/code/velox1/velox/velox/external/duckdb -isystem /usr/local/include -isystem /opt/homebrew/opt/openssl@1.1/include -D USE_VELOX_COMMON_BASE -D HAS_UNCAUGHT_EXCEPTIONS -Werror -Wall -Wextra -Wno-unused        -Wno-unused-parameter        -Wno-sign-compare        -Wno-ignored-qualifiers        -Wno-range-loop-analysis -Wno-mismatched-tags -O3 -DNDEBUG -arch arm64 -isysroot /Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk -mmacosx-version-min=12.2 -std=gnu++17 -MD -MT velox/duckdb/conversion/CMakeFiles/velox_duckdb_conversion.dir/DuckWrapper.cpp.o -MF velox/duckdb/conversion/CMakeFiles/velox_duckdb_conversion.dir/DuckWrapper.cpp.o.d -o velox/duckdb/conversion/CMakeFiles/velox_duckdb_conversion.dir/DuckWrapper.cpp.o -c /Users/jwyles/code/velox1/velox/velox/duckdb/conversion/DuckWrapper.cpp
In file included from /Users/jwyles/code/velox1/velox/velox/duckdb/conversion/DuckWrapper.cpp:16:
In file included from /Users/jwyles/code/velox1/velox/./velox/duckdb/conversion/DuckWrapper.h:20:
In file included from /Users/jwyles/code/velox1/velox/./velox/core/QueryCtx.h:24:
In file included from /Users/jwyles/code/velox1/velox/./velox/vector/DecodedVector.h:20:
In file included from /Users/jwyles/code/velox1/velox/./velox/vector/BaseVector.h:29:
In file included from /Users/jwyles/code/velox1/velox/./velox/buffer/Buffer.h:25:
In file included from /Users/jwyles/code/velox1/velox/./velox/common/base/SimdUtil.h:19:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/immintrin.h:13:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/x86gprintrin.h:15:
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/hresetintrin.h:42:27: error: invalid input constraint 'a' in asm
  __asm__ ("hreset $0" :: "a"(__eax));
                          ^
In file included from /Users/jwyles/code/velox1/velox/velox/duckdb/conversion/DuckWrapper.cpp:16:
In file included from /Users/jwyles/code/velox1/velox/./velox/duckdb/conversion/DuckWrapper.h:20:
In file included from /Users/jwyles/code/velox1/velox/./velox/core/QueryCtx.h:24:
In file included from /Users/jwyles/code/velox1/velox/./velox/vector/DecodedVector.h:20:
In file included from /Users/jwyles/code/velox1/velox/./velox/vector/BaseVector.h:29:
In file included from /Users/jwyles/code/velox1/velox/./velox/buffer/Buffer.h:25:
In file included from /Users/jwyles/code/velox1/velox/./velox/common/base/SimdUtil.h:19:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/immintrin.h:17:
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:33:5: error: use of undeclared identifier '__builtin_ia32_emms'; did you mean '__builtin_isless'?
    __builtin_ia32_emms();
    ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/c++/v1/math.h:670:12: note: '__builtin_isless' declared here
    return isless(__lcpp_x, __lcpp_y);
           ^
/Library/Developer/CommandLineTools/SDKs/MacOSX12.3.sdk/usr/include/math.h:545:22: note: expanded from macro 'isless'
#define isless(x, y) __builtin_isless((x),(y))
                     ^
In file included from /Users/jwyles/code/velox1/velox/velox/duckdb/conversion/DuckWrapper.cpp:16:
In file included from /Users/jwyles/code/velox1/velox/./velox/duckdb/conversion/DuckWrapper.h:20:
In file included from /Users/jwyles/code/velox1/velox/./velox/core/QueryCtx.h:24:
In file included from /Users/jwyles/code/velox1/velox/./velox/vector/DecodedVector.h:20:
In file included from /Users/jwyles/code/velox1/velox/./velox/vector/BaseVector.h:29:
In file included from /Users/jwyles/code/velox1/velox/./velox/buffer/Buffer.h:25:
In file included from /Users/jwyles/code/velox1/velox/./velox/common/base/SimdUtil.h:19:
In file included from /Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/immintrin.h:17:
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:33:25: error: too few arguments to function call, expected 2, have 0
    __builtin_ia32_emms();
    ~~~~~~~~~~~~~~~~~~~~^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:50:19: error: use of undeclared identifier '__builtin_ia32_vec_init_v2si'
    return (__m64)__builtin_ia32_vec_init_v2si(__i, 0);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:67:12: error: use of undeclared identifier '__builtin_ia32_vec_ext_v2si'
    return __builtin_ia32_vec_ext_v2si((__v2si)__m, 0);
           ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:129:19: error: use of undeclared identifier '__builtin_ia32_packsswb'
    return (__m64)__builtin_ia32_packsswb((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:159:19: error: use of undeclared identifier '__builtin_ia32_packssdw'
    return (__m64)__builtin_ia32_packssdw((__v2si)__m1, (__v2si)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:189:19: error: use of undeclared identifier '__builtin_ia32_packuswb'
    return (__m64)__builtin_ia32_packuswb((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:216:19: error: use of undeclared identifier '__builtin_ia32_punpckhbw'
    return (__m64)__builtin_ia32_punpckhbw((__v8qi)__m1, (__v8qi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:239:19: error: use of undeclared identifier '__builtin_ia32_punpckhwd'
    return (__m64)__builtin_ia32_punpckhwd((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:260:19: error: use of undeclared identifier '__builtin_ia32_punpckhdq'
    return (__m64)__builtin_ia32_punpckhdq((__v2si)__m1, (__v2si)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:287:19: error: use of undeclared identifier '__builtin_ia32_punpcklbw'
    return (__m64)__builtin_ia32_punpcklbw((__v8qi)__m1, (__v8qi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:310:19: error: use of undeclared identifier '__builtin_ia32_punpcklwd'
    return (__m64)__builtin_ia32_punpcklwd((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:331:19: error: use of undeclared identifier '__builtin_ia32_punpckldq'
    return (__m64)__builtin_ia32_punpckldq((__v2si)__m1, (__v2si)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:352:19: error: use of undeclared identifier '__builtin_ia32_paddb'; did you mean '__builtin_arm_addg'?
    return (__m64)__builtin_ia32_paddb((__v8qi)__m1, (__v8qi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:352:19: note: '__builtin_arm_addg' declared here
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:352:19: error: first argument of MTE builtin function must be a pointer ('__v8qi' (vector of 8 'char' values) invalid)
    return (__m64)__builtin_ia32_paddb((__v8qi)__m1, (__v8qi)__m2);
                  ^                    ~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:373:19: error: use of undeclared identifier '__builtin_ia32_paddw'; did you mean '__builtin_arm_addg'?
    return (__m64)__builtin_ia32_paddw((__v4hi)__m1, (__v4hi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:352:19: note: '__builtin_arm_addg' declared here
    return (__m64)__builtin_ia32_paddb((__v8qi)__m1, (__v8qi)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:373:19: error: first argument of MTE builtin function must be a pointer ('__v4hi' (vector of 4 'short' values) invalid)
    return (__m64)__builtin_ia32_paddw((__v4hi)__m1, (__v4hi)__m2);
                  ^                    ~~~~~~~~~~~~
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:394:19: error: use of undeclared identifier '__builtin_ia32_paddd'; did you mean '__builtin_arm_addg'?
    return (__m64)__builtin_ia32_paddd((__v2si)__m1, (__v2si)__m2);
                  ^
/Library/Developer/CommandLineTools/usr/lib/clang/13.1.6/include/mmintrin.h:352:19: note: '__builtin_arm_addg' declared here
    return (__m64)__builtin_ia32_paddb((__v8qi)__m1, (__v8qi)__m2);
                  ^
fatal error: too many errors emitted, stopping now [-ferror-limit=]
20 errors generated.
```

**Comment by yingsu00** [comment]
Looks like  `-march=native` is enabled by default. This causes issue on M1 processors. One possible way to fix this is to detect the architecture in the CMAKE file, and only add those flags on x86 machines.

**Comment by jwyles-ahana** [comment]
This is an issue because Buffer.h includes SimdUtil.h by default.

**Comment by yingsu00** [comment]
cc @majetideepak @aditi-pandit

**Comment by mbasmanova** [comment]
CC: @Yuhta @kgpai Jimmy is working on replacing explicit SIMD with xSIMD library calls.

https://github.com/xtensor-stack/xsimd#readme

**Comment by yingsu00** [comment]
Thanks @mbasmanova for the link. But it's much desired to have a non-SIMD code path as the default, since the OS can disable SIMD anyways, even on processors that support it. 
Also, I might have missed it, but AFAK, it seems the xSIMD library doesn't support all SIMD intrinsics. For example, I couldn't find its support for pdep and pext instructions that are quite essential to Varint decoding.

**Comment by mbasmanova** [comment]
Jimmy have been looking into xSIMD closely and would know better, but it is also my understanding that some instructions are missing. We were thinking it wouldn't be too difficult to contribute these to xSIMD.

**Comment by majetideepak** [comment]
The expectation to have a mvp is mid May.

**Comment by sijialouintel** [comment]
Hi @majetideepak, I am developing on ubuntu 20.04, and get the error that
![image](https://user-images.githubusercontent.com/93312725/163305203-e1dd9709-a8c1-4559-9bcb-3246fc532120.png)
Is it also caused by the compiler flags? Or, is there any solutions for it?

**Comment by yingsu00** [comment]
I'm doubtful whether using xSIMD is worthwhile. 

Firstly, its Batch<T, A> construct functionality is pretty limited(e.g. simple arithmetic operation on two batches in a loop), and for such cases, the compilers may already be able to auto-vectorize the code, even  plain code and not SIMD intrinsics.  E.g. this loop would be auto-vectorized on most recent compilers:
```
for (int i = 0; i < n; i++) {
    c[i] = a[i] + b[i];
}
```
Using xSIMD requires to rewrite the code, which itself it's a big project. Furthermore, this develops a new dependency on an external library. It's easy to add dependency than removing it and we need to be very careful.

Secondly, when we optimize for a certain architecture, we may choose different algorithms. For example, gather intrinsics is very slow on older Intel CPUs and I would avoid using it for that platform. The whole point of different compilation flags is to choose the best implementation for that platform, and for many times this means different algorithms and intrinsics. Using universal xSIMD operators/algorithms here does not help.

Thirdly, I don't know how much existing code in Velox can be transferred to use xSIMD. For example, in the previous comment __builtin_ia32_tzcnt_u64 is not supported in xSIMD yet. @Yuhta do you have some statistics how much can/need to be converted? There're many other intrinsics unsupported, meaning this library is still pre-mature. Depending on such library may add our maintenance burden in the near future.

Fourthly, @Yuhta have you done any study on the performance and code size using xSIMD? Also the compilation time? Generating code for all <T,A> combinations will increase the binary size of compilation time. With large binary size the performance may degrade because of code cache miss. If we really need to use it, we need to know this won't be an issue.

At last, using xSIMD doesn't remove the necessity of having a non-SIMD code path. Velox is open source and will be used by many users. Having a non-SIMD code path as a fall back solution would guarantee it's working for most compilers/platforms and will help get Velox promoted. 

@Yuhta Jimmy, since this is a quite important decision, do you have some time to chat about it?

**Comment by xumingming** [comment]
I run into this issue too. Would like it to be fixed, and will be happy to help.

**Comment by Yuhta** [comment]
Hi @yingsu00, sorry for the late reply.

The xSIMD work is mainly for unblocking TorchArrow release in June, which need to support Apple M1.  The current goal is just abstracting out all the `__m256` passed around, and xsimd does satisfy our needs here, otherwise we need to build our own abstraction from scratch and we probably don't have enough time for that.  On AVX2 I have verified there is no performance degradation.  Non-AVX2 platform performance is not a priority for now, but in the future we will check if we can introduce higher level dispatch to optimize for different platforms.

For BMI like `__builtin_ia32_tzcnt_u64` and `_pext_u64` we need to write our own dispatch, but there is not a lot in the code base comparing to SIMD stuff.

The work is mostly done, you can follow the first draft here in #1445.

**Comment by xumingming** [comment]
> The expectation to have an mvp is mid May.

mid May is the date when velox can build on Apple M1? @majetideepak

**Comment by Yuhta** [comment]
> > The expectation to have an mvp is mid May.
> 
> mid May is the date when velox can build on Apple M1? @majetideepak

Yes May 20 is our target date.

**Comment by yingsu00** [comment]
@Yuhta Thanks for replying. Now I see your concerns and reason why using xSIMD. I wonder if you could provide the before and after results for the following metrics before the code is merged?

- release build time
- release/debug binary size
- DWRF reader microbenchmark performance

**Comment by Yuhta** [comment]
@yingsu00 
> @Yuhta Thanks for replying. Now I see your concerns and reason why using xSIMD. I wonder if you could provide the before and after results for the following metrics before the code is merged?
> 
> * release build time
> * release/debug binary size
> * DWRF reader microbenchmark performance

I still cannot build velox with cmake (my working laptop is M1), so I am only building `velox_dwio_dwrf_reader` with buck for now as an example.  The baseline artifacts are cached in FB internal system so cannot measure the build time reliably and I am not including it here.

|Version    |Size (MB)|
|-----------|---------|
|Base (dev) |40       |
|Base (opt) |1.3      |
|New (dev)  |44       |
|New (opt)  |1.3      |

**Comment by leoYY** [comment]
Hi @Yuhta, 
I found that:
in scripts/setup-helper-functions.sh, CPU_TARGET will be setup in function "get_cxx_flags". 
But it was initialized by code above manually 
```
CPU_TARGET="${CPU_TARGET:-avx}"
```
Is there any consideration here?

**Comment by kgpai** [comment]
Hi @leoYY , 
We have details here https://github.com/facebookincubator/velox#setting-up-on-macos , on how you can get Velox running on an M1. Let us know if you run into any problems.

**Comment by ValZapod** [comment]
This is fixed in https://reviews.llvm.org/rG955cff803e081640e149fed0742f57ae1b84db7d

**Comment by Mionsz** [comment]
There can be some possible errors related when using `march=native` flag connected to AVX512 - as this seems no to be implemented yet. This can be forced using also `-mno-avx512f` flag with `-march=native`.

**Comment by majetideepak** [comment]
This is now resolved.
